test/e2e/storage/csi_mock/csi_attach_volume.go

```diff
@@ -101,7 +102,7 @@ var _ = utils.SIGDescribe("CSI Mock volume attach", func() {
 					}
 				}
 				if test.disableAttach {
-					framework.ExpectError(err, "Unexpected VolumeAttachment found")
+					gomega.Expect(err).To(gomega.HaveOccurred(), "Unexpected VolumeAttachment found")
```

Comment:
This check has a conceptual problem (with or without this change): *any* error satisfies the check, not just error that is expected by the test.

Can you use `MatchError` instead and be more precise?

This also applies to all other conversions.

Reply:
Patrick, thanks your comments. sorry for late reply. also check other conversions. will update together as soon as possible.

Reply:
Hi @pohly, the `MatchError` func doesn't accept a func as a input parameter, so the all the helper funcs in the apierrors package cannot be used in the `MatchError` func. If we want to reuse the helper funcs, we have to do like this:

```go
gomega.Expect(err).To(framework.MakeMatcher(func(actual error) (failure func() string, err error) {
	if apierrors.IsNotFound(actual) {
		return nil, nil
	}
	return func() string {
		return fmt.Sprintf("Expected VolumeAttachment not found, but got error: %s", format.Object(actual, 1))
	}, nil
}))
````

It's not very convenient and a little ugly.

IMO, if we want to resue the helper funcs, we can do like this:

```go
if !apierrors.IsNotFound(err) {
	framework.Failf("Expected VolumeAttachment not found, but got error: %s", format.Object(err, 1))
}
```

If no helper funcs we want to use and check logic is just to compare the error string, we can do like this:

```go
gomega.Expect(err).To(framework.MatchError("XXXXXX"))
```

What do you think?

Reply:
> the MatchError func doesn't accept a func as a input parameter, so the all the helper funcs in the apierrors package cannot be used in the MatchError func

That's a good point.

> `gomega.Expect(err).To(framework.MatchError("XXXXXX"))`

This already works with plain gomega: `Expect(err).To(MatchError(ContainSubstring("XXXXX"))`

Better would be some matcher that explicitly takes a `func(error) bool` and a string that explains what the function checks for:
```
Expect(err).To(BeError(apierrors.IsNotFound, "NotFound"))
```

I would make the string required, because without the failure messages are useless ("error not matched by function"). Hmm, perhaps identifying the function name could be done - then the string can be optional or be left out entirely, in favor of the usual annotations.

Another potential usage of such a matcher would be `Expect(err).To(BeError(fs.ErrExist, "ErrExist")` which then does `errors.Is(err, fs.ErrExist)` under the hood.

@onsi: do you think such a matcher would be a good addition to gomega or shall we do it instead in test/e2e/framework?

Reply:
FYI @pohly `framework.MatchError` has supported this case.

> Another potential usage of such a matcher would be Expect(err).To(BeError(fs.ErrExist, "ErrExist") which then does errors.Is(err, fs.ErrExist) under the hood.

https://github.com/carlory/kubernetes/blob/33fd0f2dddfb856e3b5daaacd1bcedf15d9f7fa3/vendor/github.com/onsi/gomega/matchers/match_error_matcher.go#L27

Reply:
True. It's just not documented (which is what I checked): https://pkg.go.dev/github.com/onsi/gomega#MatchError

Reply:
oof - clearly the docs should be updated.  adding support for a `func(error) bool` + `string` description also makes sense.  it could be a separate matcher, though as you know i'm fairly lax about changing the interfaces of these things in ways that are largely (if not precisely strictly) backward compatible.

I would lean to generalizing `MatchError(expected any)` to `MatchError(expected any, description ...any)` and allow `expected` to be of type `func(error) bool` in addition to the other types that are supported.  If `expected` _is_ of that type then `description` is required and must be a string.

We _could_ add another matcher instead... but I figured I'd float this first.

Reply:
I agree, extending `MatchError` like this is probably best. Having two different matches for errors would make it harder to find the right one.

@carlory: are you perhaps interested in providing a patch for this?

Reply:
looks like i have some time this morning.  i can work on it and ship a new version of Gomega

Reply:
i just shipped this in Gomega 1.29.0 - PTAL

Reply:
sorry for late reply.

1. update the code  and get the latest gomega version
2. check the file changed, and apply to all other conversions.
please review and comment. thanks.